src/synapse/api.py

```diff
@@ -0,0 +1,459 @@
+    """
+    res = _do_request("GET", REGISTER_URL)
+    try:
+        nonce = res.json()["nonce"]
+    except (requests.exceptions.JSONDecodeError, TypeError, KeyError) as exc:
+        logger.exception("Failed to decode nonce: %r. Received: %s", exc, res.text)
+        raise GetNonceError(str(exc)) from exc
+
+    return nonce
+
+
+def get_version() -> str:
+    """Get version.
+
+    Expected API output:
+    {
+        "server_version": "0.99.2rc1 (b=develop, abcdef123)",
+        "python_version": "3.7.8"
+    }
+
+    We're using retry here because after the config change, Synapse is restarted.
+
+    Returns:
+        The version returned by Synapse API.
+
+    Raises:
+        GetVersionError: if there was an error while reading version.
+        VersionUnexpectedContentError: if the version has unexpected content.
+    """
+    res = _do_request("GET", VERSION_URL, retry=True)
+    try:
+        server_version = res.json()["server_version"]
+    except (requests.exceptions.JSONDecodeError, KeyError, TypeError) as exc:
+        logger.exception("Failed to decode version: %r. Received: %s", exc, res.text)
+        raise GetVersionError(str(exc)) from exc
+    version_match = re.search(SYNAPSE_VERSION_REGEX, server_version)
+    if not version_match:
+        raise VersionUnexpectedContentError(
+            f"server_version has unexpected content: {server_version}"
+        )
+    return version_match.group(1)
+
+
+def get_access_token(user: User, server: str, admin_access_token: str) -> str:
+    """Get an access token that can be used to authenticate as that user.
+
+    This is a way to do actions on behalf of a user.
+
+    Args:
+        user: the user on behalf of whom you want to request the access token.
+        server: to be used to create the user id. User ID example: @user:server.com.
+        admin_access_token: a server admin access token to be used for the request.
+
+    Returns:
+        Access token.
+
+    Raises:
+        GetAccessTokenError: if there was an error while getting access token.
+    """
+    authorization_token = f"Bearer {admin_access_token}"
+    headers = {"Authorization": authorization_token}
+    # @user:server.com
+    user_id = f"@{user.username}:{server}"
+    res = _do_request("POST", f"{LOGIN_URL}/{user_id}/login", headers=headers)
+    try:
+        res_access_token = res.json()["access_token"]
+    except (requests.exceptions.JSONDecodeError, KeyError, TypeError) as exc:
+        logger.exception("Failed to decode access_token: %r. Received: %s", exc, res.text)
+        raise GetAccessTokenError(str(exc)) from exc
+    return res_access_token
+
+
+def override_rate_limit(user: User, admin_access_token: str, charm_state: CharmState) -> None:
+    """Override user's rate limit.
+
+    Args:
+        user: user to be used for requesting access token.
+        admin_access_token: server admin access token to be used.
+        charm_state: Instance of CharmState.
```

Comment:
Is this description right? It looks like this function does something other than granting access tokens

Reply:
Related: #52